libraries/AP_HAL/DSP.cpp

```diff
@@ -31,9 +34,62 @@ DSP::FFTWindowState::FFTWindowState(uint16_t window_size, uint16_t sample_rate,
 {
     // includes DC ad Nyquist components and needs to be large enough for intermediate steps
     _freq_bins = new float[window_size];
+    _hanning_window = new float[window_size];
+
+    if (_freq_bins == nullptr || _hanning_window == nullptr) {
+        delete[] _freq_bins;
+        delete[] _hanning_window;
+        _freq_bins = nullptr;
+        _hanning_window = nullptr;
+        return;
+    }
+
+    // create the Hanning window
+    // https://holometer.fnal.gov/GH_FFT.pdf - equation 19
+    for (uint16_t i = 0; i < window_size; i++) {
+        _hanning_window[i] = (0.5f - 0.5f * cosf(2.0f * M_PI * i / ((float)window_size - 1)));
+        _window_scale += _hanning_window[i];
+    }
+    // Calculate the inverse of the Effective Noise Bandwidth
+    _window_scale = 2.0f / sq(_window_scale);
 }
 
 DSP::FFTWindowState::~FFTWindowState()
 {
     delete[] _freq_bins;
+    delete[] _hanning_window;
+}
+
+// Interpolate center frequency using https://dspguru.com/dsp/howtos/how-to-interpolate-fft-peak/
+float DSP::calculate_quinns_second_estimator(FFTWindowState* fft, float* complex_fft, uint16_t k_max)
+{
+    if (k_max <= 1 || k_max >= fft->_bin_count) {
+        return 0.0f;
+    }
+
+    const uint16_t k_m1 = (k_max - 1) * 2;
+    const uint16_t k_p1 = (k_max + 1) * 2;
+    const uint16_t k = k_max * 2;
+
+    float divider = complex_fft[k] * complex_fft[k] + complex_fft[k+1] * complex_fft[k+1];
```

Comment:
those could be const
nitpick : complex_fft[k+1] style isn't the same as in the following lines

Reply:
Thanks fixed